test/widgets/test_keyboardkbdd.py

```diff
@@ -0,0 +1,157 @@
+# Copyright (c) 2021 elParaguayo
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+# Widget specific tests
+
+# The test_widget_init_config will cover the scenario where `kbdd` is not
+# running and will also run the asyncio _config_async method.
+
+# This test file covers the remaining widget code
+
+import sys
+from types import ModuleType
+
+import pytest
+
+from libqtile.bar import Bar
+
+
+def no_op(*args, **kwargs):
+    pass
+
+
+class Mockconstants(ModuleType):
+    class MessageType:
+        SIGNAL = 1
+
+
+class MockSpawn:
+    call_count = 0
+
+    @classmethod
+    def call_process(cls, *args, **kwargs):
+        if cls.call_count == 0:
+            cls.call_count += 1
+            return ""
+        else:
+            return "kbdd"
+
+
+class MockMessage:
+    def __init__(self, is_signal=True, body=0):
+        self.message_type = 1 if is_signal else 0
+        self.body = [body]
+
+
+class FakeWindow:
+    class _NestedWindow:
+        wid = 10
+    window = _NestedWindow()
+
+
+@pytest.fixture
+def patched_widget(monkeypatch):
+    monkeypatch.setitem(sys.modules, "dbus_next.constants", Mockconstants("dbus_next.constants"))
+    from libqtile.widget import keyboardkbdd
+
+    # The next line shouldn't be necessary but I got occasional failures without it when testing locally
+    monkeypatch.setattr("libqtile.widget.keyboardkbdd.MessageType", Mockconstants.MessageType)
```

Comment:
This `keyboardkbdd` test is killing me. Runs fine when running just the test file but occasionally fails when running whole `test/widgets` folder. Makes no sense (or, rather, I'm too dumb to understand why).

Patching a patched import seems to fix it but is ugly as hell!